accepted/2021/experimental-package-shading/experimental-package-shading.md

```diff
@@ -0,0 +1,339 @@
+
+We could build a mechanism for opting transitive references out of shading, but that would be _another_ mechanism and would introduce the risk of shading transitive dependencies accidentally. By making shading explicit on a per-dependency basis, we make it very clear to package authors which dependencies are being shaded.
+
+ However, direct dependencies that are shaded will unify with indirect dependencies that are transitive via other shaded dependencies. This is necessary for correctness, but it also means a package author can opt into shading a transitive dependency by making it a direct dependency.
+
+To demonstrate this, consider a project that produces a package. The project has a shaded reference to package `Foo`, and `Foo` depends on package `Bar`, so the package project transitively depends on `Bar`.
+
+```mermaid
+flowchart LR
+P[Package\nproject]
+Foo["Foo\n(shaded)"]
+Bar["Bar"]
+P --> |Shaded| Foo
+Foo --> Bar
+```
+
+Restoring, building and packing this will result in a package where the transitive reference to `Bar` from the project has become a direct dependency in the package. The shaded package `Foo` has been hidden from the dependency graph but the graph is otherwise unchanged.
+
+```mermaid
+flowchart LR
+subgraph P[Package]
+   direction TB
+   Package.dll
+   Foo.Shaded.dll
+end
+P --> Bar
+```
+
+However, if the project adds a direct shaded reference to `Bar`, this will unify with the transitive reference from `Bar`, and `Foo` will depend on the shaded version of `Bar`.
+
+```mermaid
+flowchart LR
+P[Package\nproject]
+Foo["Foo\n(shaded)"]
+Bar["Bar\n(shaded)"]
+P --> |Shaded| Foo
+P --> |Shaded| Bar
+Foo --> Bar
+```
+
+This unification is necessary so that the package project doesn't have two references to differently named copies of the same assembly, which would result in the same problems as shading public references. The package project would see multiple copies of the types from `Bar` with the exact same fully qualified names, which would cause compile errors. While it is possible to resolve this, it would substantially increase scope.
+
+Lastly, consider a case where the package project has a shaded reference to a package that's also a dependency of an unshaded reference.
+
+```mermaid
+flowchart LR
+P[Package\nproject]
+Foo["Foo"]
+Bar["Bar\n(shaded)"]
+P --> Foo
+P --> |Shaded| Bar
+Foo --> Bar
+```
+
+This is **explicitly disallowed** at this time for the same reason as the previous example: the package project's direct and transitive references to `Bar` must unify. The only way to do this is by shading `Foo` so that its references to `Bar` can be updated to target the shaded version, and we require the developer do this explicitly so that all shading operations are clear and intentional.
+
+## FAQ
```

Comment:
Not sure the right place for this but we should address the fact that re-writing an assembly is going to strip its authenticode signature and change its public key token.  This could be a problem for high security infrastructures.